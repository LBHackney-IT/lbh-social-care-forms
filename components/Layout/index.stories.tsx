import { Story } from '@storybook/react';

import Layout, { Props } from './index';
import { UserContext } from 'components/UserContext/UserContext';

export default {
  title: 'Layout',
  component: Layout,
};

const Template: Story<Props> = (args) => (
  <UserContext.Provider
    value={{
      user: {
        name: 'bar',
        hasAdminPermissions: true,
        hasDevPermissions: false,
        hasChildrenPermissions: true,
        hasAdultPermissions: true,
        email: 'foo@bar.com',
        permissionFlag: 'A',
        isAuthorised: true,
        isAuditable: false,
        isInWorkflowsPilot: false,
      },
    }}
  >
    <Layout {...args} />
  </UserContext.Provider>
);

export const Default = Template.bind({});
