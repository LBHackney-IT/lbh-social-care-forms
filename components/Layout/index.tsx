import Seo from './Seo/Seo';
import Header from './Header/Header';
import SkipLink from './SkipLink/SkipLink';
import PhaseBanner from './PhaseBanner/PhaseBanner';
import BackButton from './BackButton/BackButton';
import Footer from './Footer/Footer';
import OnboardingDialog from 'components/OnboardingDialog';
import { User } from 'types';
import { useAuth } from 'components/UserContext/UserContext';
import SiteWideAnnouncement from 'components/SitewideAnnouncement/SitewideAnnouncement';

export interface Props {
  children: React.ReactChild;
  goBackButton: boolean;
  noLayout: boolean;
}

const Layout = ({
  children,
  goBackButton = false,
  noLayout = false,
}: Props): React.ReactElement => {
  const { user } = useAuth() as { user: User };

  if (noLayout) return <>{children}</>;
  const faqLink =
    'https://docs.google.com/document/d/1eM2anY9Ddot79Gl1N386ANG6ahyn4gjZr6vsoQQGLZg/edit#';
  const handbookLink =
    'https://docs.google.com/presentation/d/1AiTljatPK58xBk2Y7R9h9mUwDpusekv2jkuCgwMWpGk/edit#slide=id.gebf6791975_1_135';

  return (
    <>
      <Seo title="Social Care Admin - Hackney Council" />
      <SkipLink />
      <Header serviceName="Social Care" />
      <PhaseBanner
        phase="support"
        faqLink={faqLink}
        handbookLink={handbookLink}
      />
      <SiteWideAnnouncement />
      {goBackButton && <BackButton />}

      <div className="govuk-width-container">
        <main className="govuk-main-wrapper" id="content" role="main">
          {children}
        </main>
      </div>

      {user?.isInWorkflowsPilot && <OnboardingDialog />}

      <Footer />
    </>
  );
};

export default Layout;
