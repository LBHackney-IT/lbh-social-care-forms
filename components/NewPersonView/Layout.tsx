import Head from 'next/head';
import Link from 'next/link';
import { format } from 'date-fns';
import { useRouter } from 'next/router';
import { Allocation, Resident, User } from 'types';
import s from './index.module.scss';
import { useRelationships } from 'utils/api/relationships';
import { useAllocatedWorkers } from 'utils/api/allocatedWorkers';
import React from 'react';
import WarningNotes from 'components/WarningNote/WarningNotes';
import { useAuth } from 'components/UserContext/UserContext';
import { canManageCases } from 'lib/permissions';
import AddFormDialog from 'components/AddFormDialog/AddFormDialog';
import { useState } from 'react';

interface NavLinkProps {
  href: string;
  children: React.ReactChild;
}

const NavLink = ({ href, children }: NavLinkProps) => {
  const router = useRouter();
  return (
    <li>
      <Link href={href}>
        <a
          className={`lbh-link lbh-link--no-visited-state ${
            router.asPath === href ? s.navLinkActive : s.navLink
          }`}
        >
          {children}
        </a>
      </Link>
    </li>
  );
};

interface Props {
  person: Resident;
  children: React.ReactChild;
}

const summariseAllocations = (allocations: Allocation[]): string | null => {
  if (allocations?.length === 1)
    return ` 路 Allocated to ${allocations[0].allocatedWorker}`;
  if (allocations?.length === 2)
    return ` 路 Allocated to ${allocations[0].allocatedWorker} and 1 other`;
  if (allocations?.length > 2)
    return ` 路 Allocated to ${allocations[0].allocatedWorker} and ${
      allocations?.length - 1
    } others`;
  return null;
};

const Layout = ({ person, children }: Props): React.ReactElement => {
  const { data: allocations } = useAllocatedWorkers(person.id);
  const { data: relationships } = useRelationships(person.id);
  const { user } = useAuth() as { user: User };

  const [addFormOpen, setAddFormOpen] = useState<boolean>(false);

  const navigation: { text: string; href: string }[] = [
    {
      text: `Allocations ${
        allocations?.allocations ? `(${allocations?.allocations?.length})` : ''
      }`,
      href: `/people/${person.id}/allocations`,
    },
    {
      text: `Relationships ${
        relationships?.personalRelationships &&
        relationships?.personalRelationships.length > 0
          ? `(${relationships?.personalRelationships.reduce(
              (count, current) => count + current.relationships.length,
              0
            )})`
          : ''
      }`,
      href: `/people/${person.id}/relationships`,
    },
    { text: 'Details', href: `/people/${person.id}/details` },
  ];

  if (canManageCases(user, person))
    navigation.unshift({
      text: 'Timeline',
      href: `/people/${person.id}`,
    });

  const secondaryNavigation = [
    {
      text: 'Add warning note',
      href: `/people/${person.id}/warning-notes/add?id=${person.id}`,
    },
  ];

  if (person.contextFlag === 'A')
    secondaryNavigation.unshift({
      text: 'Add case note',
      href: `/people/${person.id}/case-note`,
    });

  return (
    <>
      <Head>
        <title>
          {person.firstName} {person.lastName} | Social care | Hackney Council
        </title>
      </Head>

      <AddFormDialog
        isOpen={addFormOpen}
        onDismiss={() => setAddFormOpen(false)}
        person={person}
      />

      <WarningNotes id={person.id} />

      <div
        className={`govuk-grid-row govuk-!-margin-bottom-8 ${s.personHeader}`}
      >
        <div className="govuk-grid-column-two-thirds">
          <h1 className="lbh-heading-h1">
            {person.firstName} {person.lastName}
          </h1>
          <p
            className={`govuk-caption-m govuk-!-margin-top-3 ${s.personCaption}`}
          >
            #{person.id}
            {person.dateOfBirth &&
              ` 路 Born ${format(new Date(person.dateOfBirth), 'dd MMM yyyy')}`}
            {allocations?.allocations &&
              summariseAllocations(allocations.allocations)}
          </p>
        </div>

        <div className={`govuk-grid-column-one-third ${s.actionsArea}`}>
          <button
            onClick={() => setAddFormOpen(true)}
            className={`govuk-button lbh-button ${s.primaryAction}`}
          >
            <svg width="12" height="12" viewBox="0 0 12 12">
              <path d="M6.94 0L5 0V12H6.94V0Z" />
              <path d="M12 5H0V7H12V5Z" />
            </svg>
            Add something new
          </button>
        </div>
      </div>

      <div className={`govuk-grid-row ${s.outer}`}>
        <div className="govuk-grid-column-one-quarter">
          <nav className={s.sticky}>
            <ul className="lbh-list">
              {navigation.map((link) => (
                <NavLink href={link.href} key={link.href}>
                  {link.text}
                </NavLink>
              ))}
            </ul>

            <ul className={`lbh-list ${s.secondaryNav}`}>
              {secondaryNavigation.map((link) => (
                <li key={link.href}>
                  <a
                    href={link.href}
                    className="lbh-link lbh-body-s lbh-link--no-visited-state"
                  >
                    {link.text}
                  </a>
                </li>
              ))}
            </ul>
          </nav>
        </div>
        <div className="govuk-grid-column-three-quarters">{children}</div>
      </div>
    </>
  );
};

export default Layout;
