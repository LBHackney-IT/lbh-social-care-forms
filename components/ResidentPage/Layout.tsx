import Head from 'next/head';
import { format } from 'date-fns';
import { Resident } from 'types';
import s from './Layout.module.scss';
import { useAllocatedWorkers } from 'utils/api/allocatedWorkers';
import AddFormDialog from 'components/AddFormDialog/AddFormDialog';
import { useState } from 'react';
import {
  prettyAddress,
  prettyResidentName,
  summariseAllocations,
} from 'lib/formatters';
import ActivityTimeline from './ActivityTimeline';
import NavLink from './NavLink';
import Tip from 'components/Tip/Tip';
import StatusTags from './StatusTags';
import WarningNotes from './WarningNotes';

interface Props {
  resident: Resident;
  children: React.ReactChild | React.ReactChild[];
  title?: string;
}

const Layout = ({ title, resident, children }: Props): React.ReactElement => {
  const { data: allocations } = useAllocatedWorkers(resident.id);
  const [addFormOpen, setAddFormOpen] = useState<boolean>(false);

  const navigation = [
    { text: 'Overview', href: `/residents/${resident.id}` },
    {
      text: 'Workflows',
      href: `/residents/${resident.id}/workflows`,
      tip: "A timeline of the resident's care, assessments and support plans",
    },
    {
      text: 'Case notes & records',
      href: `/residents/${resident.id}/case-notes`,
      tip: 'You can also find old Google form submissions here',
    },
    {
      text: `Allocations`,
      href: `/residents/${resident.id}/allocations`,
    },
    { text: 'Relationships', href: `/residents/${resident.id}/relationships` },
  ];

  const secondaryNavigation = [
    // {
    //   text: 'Shareable version',
    //   href: `/people/${resident.id}/warning-notes/add?id=${resident.id}`,
    //   tip: "A handy printer-friendly view of this resident's info",
    // },
    {
      text: 'Add case note',
      href: `/people/${resident.id}/case-note`,
    },
  ];

  return (
    <>
      <Head>
        <title>
          {title && `${title} | `}
          {prettyResidentName(resident)} | Social care | Hackney Council
        </title>
      </Head>

      <AddFormDialog
        isOpen={addFormOpen}
        onDismiss={() => setAddFormOpen(false)}
        person={resident}
      />

      <div
        className={`govuk-grid-row govuk-!-margin-bottom-8 ${s.residentHeader}`}
      >
        <div className={`govuk-grid-column-two-thirds ${s.masthead}`}>
          <h1>{prettyResidentName(resident)}</h1>

          <p className="lbh-body-s">
            #{resident.id}
            {resident.dateOfBirth &&
              ` · Born ${format(new Date(resident.dateOfBirth), 'd MMM yyyy')}`}
            {resident.address && ` · ${prettyAddress(resident)}`}
          </p>

          {allocations?.allocations && (
            <p className="lbh-body-xs">
              {summariseAllocations(allocations.allocations)}
              {resident.allocatedTeam && ` · ${resident.allocatedTeam}`}
            </p>
          )}

          <StatusTags resident={resident} allocations={allocations} />

          {/* <CaseStatusFlag person={resident} /> */}
        </div>

        <div className={`govuk-grid-column-one-third ${s.actionsArea}`}>
          <Tip content="Add case notes, assessments, warning notes or other records">
            <button
              onClick={() => setAddFormOpen(true)}
              className={`govuk-button lbh-button ${s.primaryAction}`}
            >
              <svg width="12" height="12" viewBox="0 0 12 12">
                <path d="M6.94 0L5 0V12H6.94V0Z" />
                <path d="M12 5H0V7H12V5Z" />
              </svg>
              Add something new
            </button>
          </Tip>
        </div>
      </div>

      <div className={`govuk-grid-row ${s.outer}`}>
        <div className="govuk-grid-column-one-quarter">
          <nav className={s.sticky}>
            <ul className="lbh-list">
              {navigation.map((link) => (
                <NavLink href={link.href} key={link.href} tip={link.tip}>
                  {link.text}
                </NavLink>
              ))}
            </ul>

            <ul className={`lbh-list ${s.secondaryNav}`}>
              {secondaryNavigation.map((link) => (
                <NavLink
                  href={link.href}
                  key={link.href}
                  tip={link.tip}
                  secondary
                >
                  {link.text}
                </NavLink>
              ))}
            </ul>

            <ActivityTimeline socialCareId={resident.id} />
          </nav>
        </div>

        <div className="govuk-grid-column-three-quarters">
          <WarningNotes socialCareId={resident.id} />

          {children}
        </div>
      </div>
    </>
  );
};

export default Layout;
