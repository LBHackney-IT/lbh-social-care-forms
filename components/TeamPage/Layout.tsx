import classNames from 'classnames';
import Link from 'next/link';
import { useRouter } from 'next/router';
import { Team } from 'types';
import { useTeamWorkers } from 'utils/api/allocatedWorkers';

interface NavLinkProps {
  href: string;
  children: React.ReactChild;
}

interface Props {
  team?: Team;
  children: React.ReactChild;
}

const NavLink = ({ href, children }: NavLinkProps) => {
  const router = useRouter();

  return (
    <li
      className={classNames(
        'govuk-tabs__list-item',
        router.asPath === href && 'govuk-tabs__list-item--selected'
      )}
    >
      <Link href={href} replace={true} scroll={false}>
        <a className="govuk-tabs__tab" data-testid={`navlink_content_${href}`}>
          {children}
        </a>
      </Link>
    </li>
  );
};

const TeamLayout = ({ team, children }: Props): React.ReactElement => {
  const { data: users } = useTeamWorkers(team?.id);

  return (
    <>
      <p className="govuk-caption-s">Team</p>
      <h1 className="govuk-!-margin-top-1">{team?.name || 'Unknown team'}</h1>

      {team && (
        <div className="govuk-tabs lbh-tabs govuk-!-margin-top-8">
          <ul className="govuk-tabs__list">
            <NavLink href={`/teams/${team.id}`}>Waiting list</NavLink>
            <NavLink href={`/teams/${team.id}/members`}>
              <>Members {users?.length && `(${users.length})`}</>
            </NavLink>
            <NavLink href={`/teams/${team.id}/allocations`}>
              Allocations
            </NavLink>
          </ul>

          <section className="govuk-tabs__panel" id="past-day">
            {children}
          </section>
        </div>
      )}
    </>
  );
};
export default TeamLayout;
