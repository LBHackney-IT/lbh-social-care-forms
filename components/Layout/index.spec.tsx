import { render, screen } from '@testing-library/react';
import {
  FeatureFlagProvider,
  FeatureSet,
} from '../../lib/feature-flags/feature-flags';

import Layout from './index';

jest.mock('next/router', () => ({
  useRouter: () => ({
    asPath: 'path',
  }),
}));

jest.mock('./Header/Header', () => () => 'Header');
jest.mock('./PhaseBanner/PhaseBanner', () => () => 'MockedPhaseBanner');

const features: FeatureSet = {
  'workflows-pilot': {
    isActive: false,
  },
};

describe('Layout component', () => {
  it('should render properly', async () => {
    const { asFragment } = render(
      <FeatureFlagProvider features={features}>
        <Layout goBackButton={false} noLayout={false}>
          <p>I am the children</p>
        </Layout>
      </FeatureFlagProvider>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  describe('when workflows pilot feature flag is on', () => {
    it('displays the onboarding dialog', () => {
      render(
        <FeatureFlagProvider
          features={{
            'workflows-pilot': {
              isActive: true,
            },
          }}
        >
          <Layout goBackButton={false} noLayout={false}>
            <p>I am the children</p>
          </Layout>
        </FeatureFlagProvider>
      );

      expect(
        screen.queryByText(
          "Good news! You're part of the brand new workflow pilot."
        )
      ).toBeVisible();
    });
  });

  describe('when workflows pilot feature flag is off', () => {
    it("doesn't display the onboarding dialog", () => {
      render(
        <FeatureFlagProvider
          features={{
            'workflows-pilot': {
              isActive: false,
            },
          }}
        >
          <Layout goBackButton={false} noLayout={false}>
            <p>I am the children</p>
          </Layout>
        </FeatureFlagProvider>
      );

      expect(
        screen.queryByText(
          "Good news! You're part of the brand new workflow pilot."
        )
      ).not.toBeInTheDocument();
    });
  });
});
