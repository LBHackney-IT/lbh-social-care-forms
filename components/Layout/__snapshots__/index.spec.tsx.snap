// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout component should render properly 1`] = `
<DocumentFragment>
  <a
    class="govuk-skip-link"
    href="#content"
  >
    Skip to main content
  </a>
  HeaderMockedPhaseBanner
  <section
    class="lbh-announcement lbh-announcement--site lbh-announcement--site--warning"
  >
    <div
      class="lbh-container"
    >
      <h3
        class="lbh-announcement__title"
      >
        DO 
        <span
          class="underline"
        >
          NOT
        </span>
         RECORD HERE !
      </h3>
      <div
        class="lbh-announcement__content"
      >
        <p>
          All information needs to be entered onto Mosaic only.
          <br />
          See the 
          <a
            href="https://intranet.hackney.gov.uk/mosaic-for-adults-social-care"
          >
             Adults Social Care Mosaic Pages 
          </a>
           on the intranet
        </p>
      </div>
    </div>
  </section>
  <div
    class="govuk-width-container"
  >
    <main
      class="govuk-main-wrapper"
      id="content"
      role="main"
    >
      <p>
        I am the children
      </p>
    </main>
  </div>
  <footer
    class="footer"
    role="contentinfo"
  >
    <div
      class="govuk-width-container"
    >
      <nav
        class="links"
      >
        <a
          href="https://sites.google.com/hackney.gov.uk/moderntoolsforsocialcare"
        >
          Roadmap
        </a>
        <a
          href="https://docs.google.com/forms/d/e/1FAIpQLScILbPD1ioKHzp1D3HN4_DKaxV2tpWLMu8upSSqNgSPCo85cg/viewform"
        >
          Give feedback
        </a>
      </nav>
      <div
        class="meta"
      >
        Built and maintained by HackIT.
      </div>
    </div>
  </footer>
</DocumentFragment>
`;
